### DISCUSSION TITLE
Supporting installation as a Progressive Web App

### DISCUSSION THREAD

**Discussion by agdturner** [molstar/molstar#1390] [Ideas]
Hi and thanks for Mol* :)

I work on an academic research project (https://ribocode.org/) from January to August 2025. The bit of the project I am helping with is developing a tool to visualise ribosomes. I forked Mol*, activated GitHub pages and did some changes to have Mol* installable as a Progressive Web App (PWA). See references below for details of what a PWA is.

It was not many changes for a PWA, and I thought it might be something generally wanted? If so, I am happy to contribute those changes and work with someone to ensure the install will work for the Mol* main site and so that it is easy for others to spin up their own. Here is a link to my forked repository:
https://github.com/agdturner/molstar/
Here is a link to the GitHub Pages for the viewer:
 - https://agdturner.github.io/molstar/build/viewer

I have found installation as a PWA easiest with Chrome. From the above viewer link in Chrome, you should see a button in the right of the URL bar that can be actioned to install this version of Mol* as a PWA.

Changes made:
- src/apps/viewer/index.html
  - manifest added in the HTML <head> (link to a new manifest.webmanifest file)
  - At the end of the HTML <body> the service worker is registered
- sw.js and manifest.webmanifest files added in the root directory.

The benefits of a PWA is that it is relatively easy for users to install and use offline with their own data.

Would you like this contributing? If so, how to test it and configure things? I imagine there could be a configuration step wanted in the build process to ensure the link in the viewer html points to the right place. Also using a variable for the version, the version of the PWA cache could also be updated... 

Best wishes,

Andy

References
1. https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps
2. https://en.wikipedia.org/wiki/Progressive_web_app)

**Comment by dsehnal** [comment]
Hi, thanks for reaching out. This seems like a valuable contribution. Do you see a way how we could add this as an optional feature for users? Perhaps adding it as an example or a separate app (or just a separate `index-pwa.html`?

**Reply by agdturner** [reply]
Thanks for getting back to us David.

The next things I was thinking to try:
1. Have different service workers for each of the different index.html files created in the build directory.
2. Understanding more about the build process with a view to injecting the necessary things into the index.html files as part of the build process.

I don't think separate files would be wanted in the end, but I can try creating additional index-pwa.html files for the time being.

**Reply by agdturner** [reply]
I created code for having different PWAs for viewer, docking-viewer, mesoscale-explorer and the examples. There are a couple of scripts (pwa.js and pwa-clean.js) to automate the code generation (pwa.js) and for cleaning up (pwa-clean.js). For each PWA, the index-pwa.html is generated from the original index.html. The respective index.ts and index.tsx files were updated to add reference to these additional html files. The service worker and manifest files are (re)generated as part of the build.

The PWAs can be found here:
https://agdturner.github.io/molstar/build/viewer/index-pwa.html
https://agdturner.github.io/molstar/build/docking-viewer/index-pwa.html
https://agdturner.github.io/molstar/build/mesoscale-explorer/index-pwa.html
https://agdturner.github.io/molstar/build/examples/alpha-orbitals/index-pwa.html
https://agdturner.github.io/molstar/build/examples/basic-wrapper/index-pwa.html
https://agdturner.github.io/molstar/build/examples/lighting/index-pwa.html
https://agdturner.github.io/molstar/build/examples/proteopedia-wrapper/index-pwa.html

What are the next steps @dsehnal ?

**Reply by agdturner** [reply]
I have submitted a PR: https://github.com/molstar/molstar/pull/1416/
